### PR TITLE
Fix call context propagation

### LINKED ISSUE #177: Call contexts aren't propagated properly
Originally reported by: **Claudiu Popa (BitBucket: [PCManticore](http://bitbucket.org/PCManticore), GitHub: @PCManticore)**

---

Given the following code, astroid can't understand it, because call contexts are improperly passed in the inference chain. 

In this particular example, when chest is called, its call context will contain the arguments
Name(a), Name(b), without the references to their original values.

```
#!python

from astroid.test_utils import extract_node
n = extract_node('''
def chest(a):
    return a * a

def best(a, b):
    return chest(a)

def test(a, b, c):
    return best(a, b)

test(4, 5, 6) #@
''')
print(n)
print(n.infered())
```

---
- Bitbucket: https://bitbucket.org/logilab/astroid/issue/177

### PR DESCRIPTION
**Pull request by brycepg**
Fix inference of call arguments

Close #177
Close #404 
Close PyCQA/pylint#224
Close PyCQA/pylint#1660

### CODE REVIEW

**Comment by brycepg** [comment]
Note: `pytest -k test_numpy_crash` took multiple minutes to complete for me locally.

**Comment by brycepg** [comment]
I may have to do some performance tuning before this gets accepted

**Comment by brycepg** [comment]
A problem I'm seeing is that arguments with multiple inferences cause the function result to become `Uninferable`. Do you think I should cover this case?

**Comment by brycepg** [comment]
I think I'm going to try to change nodes to allow an explicit context, which will override regular inference contexts to avoid Inference too soon

**Comment by brycepg** [comment]
This is causing a recursion error in home-assistant/home-assistant. I'll have to fix that
Integration tests are already paying off :)

**Comment by brycepg** [comment]
I think I'm going to rewrite this

File: astroid/inference.py [hunk 1]
```diff
@@ -518,6 +519,8 @@ def _is_not_implemented(const):
 def _invoke_binop_inference(instance, opnode, op, other, context, method_name):
     """Invoke binary operation inference on the given instance."""
     methods = dunder_lookup.lookup(instance, method_name)
+    if context is not None:
```
Comment (PCManticore) [inline astroid/inference.py]: I think this could definitely be in another PR.

File: astroid/inference.py [hunk 2]
```diff
@@ -832,3 +835,17 @@ def instance_getitem(self, index, context=None):
             node=self, index=index, context=context))
 
 bases.Instance.getitem = instance_getitem
+
+
+def _set_explicit_context_for_call(call, context):
```
Comment (PCManticore) [inline astroid/inference.py]: One problem I have with this is that it mutates the underlying nodes with a context. While we do that for the explicit inference, I'd like us to start considering the nodes as immutable, with as few as possible places where we'd have to mutate them (right now another one apart of `explicit_inference` would be mutating `.locals`). It would make more sense to push these values inside the context itself, after which we could modify the inference to look for the nested values of a context when inferring values. This way we keep nodes untouched, we only pass the modified context.

File: astroid/inference.py (hunk 1 above)
Comment (brycepg) [reply astroid/inference.py]: I'll see if this passes any of the tests in a new PR. I think it depends on the explicit context

**Comment by PCManticore** [comment]
Closed as we already merged the improvement based on a new parameter.
